### PR TITLE
Implement remote Api and work on character UI

### PR DESCRIPTION
**Pull request by sahar-as**
Report: 
- Implement get Character API
- Implement Rick & Morty Repository and its remote implementation
- Create Rick & Morty use a case in order to convert massive data to a Character Model (id, name, image)
- Make the character UI part more completed

### CODE REVIEW

File: app/src/main/java/ir/saharapps/rickandmorty/data/di/NetworkModule.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+@Module
+@InstallIn(SingletonComponent::class)
+object NetworkModule {
+
+    @Singleton
+    @Provides
+    fun provideRickMortyRetrofit(): RickAndMortyApi{
+        return Retrofit.Builder()
+            .baseUrl(RICK_AND_MORTY_URL)
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/data/di/NetworkModule.kt:21]: `Constants.RICK_AND_MORTY_URL`
APIs.BASE_URL

File: app/src/main/java/ir/saharapps/rickandmorty/data/dto/CharacterApiResult.kt [hunk 2]
```diff
@@ -1,6 +1,6 @@
 package ir.saharapps.rickandmorty.data.dto
 
-data class RickAndMortyDto(
+data class CharacterApiResult(
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/data/dto/CharacterApiResult.kt:3]: Kotlin serialization @Keep

File: app/src/main/java/ir/saharapps/rickandmorty/data/remote/RickAndMortyApi.kt [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
 
-import ir.saharapps.rickandmorty.data.dto.RickAndMortyDto
+import ir.saharapps.rickandmorty.data.dto.CharacterApiResult
 import retrofit2.http.GET
 
 interface RickAndMortyApi {
 
-    @GET("character")
-    suspend fun getCharacter(): RickAndMortyDto
+    @GET("character/")
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/data/remote/RickAndMortyApi.kt:8]: `APIs.charachers ...`

File: app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRemoteRepositoryImpl.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package ir.saharapps.rickandmorty.data.repository
+
+import ir.saharapps.rickandmorty.data.dto.CharacterApiResult
+import ir.saharapps.rickandmorty.data.dto.Result
+import ir.saharapps.rickandmorty.data.remote.RickAndMortyApi
+import javax.inject.Inject
+
+class RickMortyRemoteRepositoryImpl @Inject constructor(
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRemoteRepositoryImpl.kt:8]: Rename `RickMortyRemoteRepositoryImpl` to better name

File: app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRepository.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package ir.saharapps.rickandmorty.data.repository
+
+import ir.saharapps.rickandmorty.data.dto.Result
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRepository.kt:3]: Rename `RickMortyRepository` to better name

File: app/src/main/java/ir/saharapps/rickandmorty/domain/usecase/RickMortyRemoteUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package ir.saharapps.rickandmorty.domain.usecase
+
+import ir.saharapps.rickandmorty.data.repository.RickMortyRemoteRepositoryImpl
+import ir.saharapps.rickandmorty.domain.model.Character
+import javax.inject.Inject
+
+class RickMortyRemoteUseCase @Inject constructor(
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/domain/usecase/RickMortyRemoteUseCase.kt:7]: Rename to better name

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt [hunk 7]
```diff
@@ -5,37 +5,40 @@ import android.view.ViewGroup
...
         holder.bind(currentItem)
     }
 
-    inner class RickMortyViewHolder(private val binding: CharacterItemBinding): RecyclerView.ViewHolder(binding.root){
-        fun bind(ricKAndMorty: RicKAndMorty){
-            //Todo assign image with glide
-            binding.txtCharacterName.text = ricKAndMorty.name
+    inner class CharactersViewHolder(private val binding: ItemCharacterBinding): RecyclerView.ViewHolder(binding.root){
+        fun bind(character: Character){
+            Glide.with(binding.root.context)
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt:26]: It's work and fine but you can write a extension function and reusable code

**Review by FarshidRoohi — CHANGES_REQUESTED** [review]

File: app/src/main/java/ir/saharapps/rickandmorty/data/di/NetworkModule.kt (hunk 1 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/data/di/NetworkModule.kt:21]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/data/dto/CharacterApiResult.kt (hunk 2 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/data/dto/CharacterApiResult.kt:3]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/data/remote/RickAndMortyApi.kt (hunk 3 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/data/remote/RickAndMortyApi.kt:8]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRemoteRepositoryImpl.kt (hunk 4 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRemoteRepositoryImpl.kt:8]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRepository.kt (hunk 5 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/data/repository/RickMortyRepository.kt:3]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/domain/usecase/RickMortyRemoteUseCase.kt (hunk 6 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/domain/usecase/RickMortyRemoteUseCase.kt:7]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt (hunk 7 above)
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt:26]: Done

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt [hunk 8]
```diff
@@ -5,37 +5,42 @@ import android.view.ViewGroup
...
         holder.bind(currentItem)
     }
 
-    inner class RickMortyViewHolder(private val binding: CharacterItemBinding): RecyclerView.ViewHolder(binding.root){
-        fun bind(ricKAndMorty: RicKAndMorty){
-            //Todo assign image with glide
-            binding.txtCharacterName.text = ricKAndMorty.name
+    inner class CharactersViewHolder(private val binding: ItemCharacterBinding): RecyclerView.ViewHolder(binding.root){
+        fun bind(character: Character){
+//            Glide.with(binding.root.context)
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt:26]: first one please remove comments and you can write better name for load image and must not be a name contains extension

**Review by FarshidRoohi — CHANGES_REQUESTED** [review]

**Review by FarshidRoohi — CHANGES_REQUESTED** [review]

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt [hunk 9]
```diff
@@ -5,37 +5,38 @@ import android.view.ViewGroup
...
         holder.bind(currentItem)
     }
 
-    inner class RickMortyViewHolder(private val binding: CharacterItemBinding): RecyclerView.ViewHolder(binding.root){
-        fun bind(ricKAndMorty: RicKAndMorty){
-            //Todo assign image with glide
-            binding.txtCharacterName.text = ricKAndMorty.name
+    inner class CharactersViewHolder(private val binding: ItemCharacterBinding): RecyclerView.ViewHolder(binding.root){
+        fun bind(character: Character){
+            binding.imgCharacterPic.setImage(character.image)
```
[the remark this answers is not in the document]
Comment (sahar-as) [reply app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters_screen/CharactersAdapter.kt]: Done

**Pull request merged**